src/util/audio/PortAudioProducer.cpp

```diff
@@ -5,123 +5,160 @@ PortAudioProducer::PortAudioProducer(Settings *settings, AudioQueue *audioQueue)
    settings(settings),
    audioQueue(audioQueue)
 {
-	XOJ_INIT_TYPE(PortAudioProducer);
-
-	DeviceInfo inputInfo(&sys.defaultInputDevice(), true);
-	this->setInputDevice(inputInfo);
+    XOJ_INIT_TYPE(PortAudioProducer);
 }
 
 PortAudioProducer::~PortAudioProducer()
 {
-	XOJ_CHECK_TYPE(PortAudioProducer);
+    XOJ_CHECK_TYPE(PortAudioProducer);
 
-	portaudio::System::terminate();
+    if (portaudio::System::exists())
+    {
+        portaudio::System::terminate();
+    }
 
-	XOJ_RELEASE_TYPE(PortAudioProducer);
+    XOJ_RELEASE_TYPE(PortAudioProducer);
 }
 
 std::list<DeviceInfo> PortAudioProducer::getInputDevices()
 {
-	XOJ_CHECK_TYPE(PortAudioProducer);
+    XOJ_CHECK_TYPE(PortAudioProducer);
 
-	std::list<DeviceInfo> deviceList;
+    std::list<DeviceInfo> deviceList;
 
-	for (portaudio::System::DeviceIterator i = this->sys.devicesBegin(); i != sys.devicesEnd(); ++i)
-	{
+    for (portaudio::System::DeviceIterator i = this->sys.devicesBegin(); i != sys.devicesEnd(); ++i)
+    {
 
-		if (i->isFullDuplexDevice() || i->isInputOnlyDevice())
-		{
-			DeviceInfo deviceInfo(&(*i), this->selectedInputDevice == i->index());
-			deviceList.push_back(deviceInfo);
-		}
+        if (i->isFullDuplexDevice() || i->isInputOnlyDevice())
+        {
+            DeviceInfo deviceInfo(&(*i), this->settings->getAudioInputDevice() == i->index());
+            deviceList.push_back(deviceInfo);
+        }
 
-	}
-	return deviceList;
+    }
+    return deviceList;
 }
 
 const DeviceInfo PortAudioProducer::getSelectedInputDevice()
 {
-	XOJ_CHECK_TYPE(PortAudioProducer);
-
-	return DeviceInfo(&sys.deviceByIndex(this->selectedInputDevice), true);
-}
-
-void PortAudioProducer::setInputDevice(DeviceInfo deviceInfo)
-{
-	XOJ_CHECK_TYPE(PortAudioProducer);
-
-	this->selectedInputDevice = deviceInfo.getIndex();
-	portaudio::Device *device = &sys.deviceByIndex(this->selectedInputDevice);
-	this->inputChannels = static_cast<unsigned int>(device->maxInputChannels());
+    XOJ_CHECK_TYPE(PortAudioProducer);
+
+    try
+    {
+        return DeviceInfo(&sys.deviceByIndex(this->settings->getAudioInputDevice()), true);
+    }
+    catch(portaudio::PaException &e)
+    {
+        g_message("PortAudioProducer: Selected input device not found - fallback to default input device");
```

Comment:
May add e.cause to the log, as there may another exception occur

Reply:
There is only one cause for this exception: The device, selected in the settings, got disconnected or is for some other reason not available.

Should I add a more verbose description or do you mean some special command I have not yet seen?

Reply:
you should log e.what(), because you never know, may the next version provide a string or so, so usually I would log it, a normal user don't see it anyway. (add it to the existing message, nothing special)